lib/ansible/modules/crypto/openssl_privatekey_info.py

```diff
@@ -0,0 +1,462 @@
+                    # Get public and private key exponents
+                    y = OpenSSL._util.ffi.new("BIGNUM **")
+                    x = OpenSSL._util.ffi.new("BIGNUM **")
+                    OpenSSL._util.lib.DSA_get0_key(key, y, x)
+                    key_public_data['y'] = self.bigint_to_int(y[0])
+                    key_private_data['x'] = self.bigint_to_int(x[0])
+                else:
+                    # Get public parameters (primes and group element)
+                    key_public_data['p'] = self.bigint_to_int(key.p)
+                    key_public_data['q'] = self.bigint_to_int(key.q)
+                    key_public_data['g'] = self.bigint_to_int(key.g)
+                    # Get public and private key exponents
+                    key_public_data['y'] = self.bigint_to_int(key.pub_key)
+                    key_private_data['x'] = self.bigint_to_int(key.priv_key)
+                try_fallback = False
+            except AttributeError:
+                # Use fallback if available
+                pass
+        else:
+            # Return 'unknown'
+            key_type = 'unknown ({0})'.format(self.key.type())
+        # If needed and if possible, fall back to cryptography
+        if try_fallback and PYOPENSSL_VERSION >= LooseVersion('16.1.0') and CRYPTOGRAPHY_FOUND:
+            return _get_cryptography_key_info(self.key.to_cryptography_key())
+        return key_type, key_public_data, key_private_data
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            path=dict(type='path', required=True),
+            passphrase=dict(type='str', no_log=True),
+            return_private_key_data=dict(type='bool', default=False),
+            select_crypto_backend=dict(type='str', default='auto', choices=['auto', 'cryptography', 'pyopenssl']),
+        ),
+        supports_check_mode=True,
+    )
+
+    try:
+        base_dir = os.path.dirname(module.params['path']) or '.'
+        if not os.path.isdir(base_dir):
+            module.fail_json(
+                name=base_dir,
+                msg='The directory %s does not exist or the file is not a directory' % base_dir
+            )
+
+        backend = module.params['select_crypto_backend']
+        if backend == 'auto':
+            # Detect what backend we can use
+            can_use_cryptography = CRYPTOGRAPHY_FOUND and CRYPTOGRAPHY_VERSION >= LooseVersion(MINIMAL_CRYPTOGRAPHY_VERSION)
+            can_use_pyopenssl = PYOPENSSL_FOUND and PYOPENSSL_VERSION >= LooseVersion(MINIMAL_PYOPENSSL_VERSION)
+
+            # If cryptography is available we'll use it
+            if can_use_cryptography:
+                backend = 'cryptography'
+            elif can_use_pyopenssl:
+                backend = 'pyopenssl'
+
+            # Fail if no backend has been found
+            if backend == 'auto':
+                module.fail_json(msg=("Can't detect none of the required Python libraries "
```

Comment:
Should be "Can't detect any" rather than "none" I think.

Reply:
Fixed (also in other modules with same/similar formulation).